dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/telemetry/WithSpanServlet.java

```diff
@@ -0,0 +1,115 @@
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.instrumentation.annotations.SpanAttribute;
+import io.opentelemetry.instrumentation.annotations.WithSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.servlet.annotation.WebServlet;
+
+@SuppressWarnings("serial")
+@WebServlet("/WithSpanServlet")
+public class WithSpanServlet extends FATServlet {
+
+    @Inject
+    private SpanBean spanBean;
+
+    private static final String PARAMETER_TEST = "testString";
+
+    @Test
+    public void callNotAnnotated() {
+        //Returned spanId from methodNotAnnotated should not be the spanID of 0000000000000000
+        //No span was created
+        String spanId = spanBean.methodNotAnnotated();
+        assertThat(spanId, equalTo("0000000000000000"));
+    }
+
+    @Test
+    public void callAnnotated() {
+        String spanId = spanBean.methodAnnotated();
+        //Span should end when returned to this method so the current span should have the default spanID of 0000000000000000
+        Span span = Span.current();
+        assertThat(span.getSpanContext().getSpanId(), equalTo("0000000000000000"));
+
+        //Returned spanId from methodAnnotated should not be the default spanID
+        assertThat(spanId, not(equalTo("0000000000000000")));
+    }
+
+    @Test
+    public void callMethodAnnotatedWithParameter() {
+        String spanParameter = spanBean.methodAnnotatedWithParameter(PARAMETER_TEST);
+        //Span should end when returned to this method so the current span should have the default spanID of 0000000000000000
+        assertThat(spanParameter, equalTo(PARAMETER_TEST));
+    }
+
+    /*
+     * Make multiple calls to
+     *
+     * @Test
+     * public void callRecursiveAnnotated() {
+     * int calls = 2;
+     * Span span = Span.current();
+     * String previousId = span.getSpanContext().getSpanId();
+     * spanBean.methodRecursiveAnnotated(calls, previousId);
+     * }
+     */
+
+    @ApplicationScoped
+    public static class SpanBean {
+
+        //Creates a span for this method
+        @WithSpan
+        String methodAnnotated() {
+            Span span = Span.current();
+            return span.getSpanContext().getSpanId();
+        }
+
+        //Creates a span for this method
+        //Method calls itself a number of times and compares the current span with the parent span
+        //TO DO
+        //The currentId remains the same as the previousId - It's not clear if this is expected behaviour or not
+        @WithSpan
+        void methodRecursiveAnnotated(int calls, String previousId) {
+            Span span = Span.current();
+            String currentId = span.getSpanContext().getSpanId();
+            System.out.println(currentId + " " + previousId);
+            if (calls > 0) {
+                methodRecursiveAnnotated(calls - 1, currentId);
+            }
+        }
```

Comment:
The reason this isn't working is that CDI can only call interceptors when you call the method on an object that has been injected by CDI.

When you're calling from outside the bean, you're generally doing this anyway and don't really think about it.

However, if you're calling a method on the same bean and you want interceptors to fire, you need to inject the bean into itself and call the method on the injected object.

E.g.
```java
    @ApplicationScoped
    public static class SpanBean {
        
        @Inject
        private SpanBean thisBean;
        
        @WithSpan
        void methodRecursiveAnnotated(int calls, String previousId) {
            Span span = Span.current();
            String currentId = span.getSpanContext().getSpanId();
            System.out.println(currentId + " " + previousId);
            if (calls > 0) {
                thisBean.methodRecursiveAnnotated(calls - 1, currentId);
            }
        }

    }
```

Reply:
Also, for this method you might want to pass a list into the method and have each call append its spanId to the list. Then you can assert that all the SpanIds in the list are different.